src/capture/capture.cpp

```diff
@@ -23,141 +23,372 @@
+{
+	switch (type) {
+	case CaptureType::Audio: return ".wav";
+	case CaptureType::Midi: return ".mid";
+	case CaptureType::RawOplStream: return ".dro";
+	case CaptureType::RadOplInstruments: return ".rad";
+
+	case CaptureType::Video: return ".avi";
+
+	case CaptureType::RawImage:
+	case CaptureType::UpscaledImage:
+	case CaptureType::RenderedImage: return ".png";
+
+	case CaptureType::SerialLog: return ".serlog";
+	}
+}
+
+static std::string capture_type_to_postfix(const CaptureType type)
+{
+	switch (type) {
+	case CaptureType::RawImage: return "-raw";
+	case CaptureType::UpscaledImage: return "";
+	case CaptureType::RenderedImage: return "-rendered";
+	default: return "";
+	}
+}
+
+static int32_t get_next_capture_index(const CaptureType type)
+{
+	switch (type) {
+	case CaptureType::Audio: return capture.next_index.audio;
+	case CaptureType::Midi: return capture.next_index.midi;
+
+	case CaptureType::RawOplStream:
+		return capture.next_index.raw_opl_stream;
+
+	case CaptureType::RadOplInstruments:
+		return capture.next_index.rad_opl_instrument;
+
+	case CaptureType::Video: return capture.next_index.video;
+
+	case CaptureType::RawImage:
+	case CaptureType::UpscaledImage:
+	case CaptureType::RenderedImage: return capture.next_index.image;
+
+	case CaptureType::SerialLog: return capture.next_index.serial_log;
+	}
+}
+
+static void set_next_capture_index(const CaptureType type, int32_t index)
+{
+	switch (type) {
+	case CaptureType::Audio: capture.next_index.audio = index; break;
+	case CaptureType::Midi: capture.next_index.midi = index; break;
+
+	case CaptureType::RawOplStream:
+		capture.next_index.raw_opl_stream = index;
+		break;
+
+	case CaptureType::RadOplInstruments:
+		capture.next_index.rad_opl_instrument = index;
+		break;
+
+	case CaptureType::Video: capture.next_index.video = index; break;
+
+	case CaptureType::RawImage:
+	case CaptureType::UpscaledImage:
+	case CaptureType::RenderedImage:
+		capture.next_index.image = index;
+		break;
+
+	case CaptureType::SerialLog:
+		capture.next_index.serial_log = index;
+		break;
+	}
+}
+
+#pragma warning( pop )
+
+static bool maybe_create_capture_directory(const CaptureType type)
+{
+	static std::mutex mutex;
+
+	// As the various capturers might be running in separate threads (e.g.
+	// the image capturer uses multiple threads) and the filesystem is a
+	// shared resource, creating the output directory is prone to race
+	// conditions.
+	std::lock_guard<std::mutex> guard(mutex);
+
+	std::error_code ec = {};
+
+	if (!std_fs::exists(capture.path, ec)) {
+		if (!std_fs::create_directory(capture.path, ec)) {
+			LOG_WARNING("CAPTURE: Can't create directory '%s' for capturing %s: %s",
+			            capture.path.c_str(),
+			            capture_type_to_string(type).c_str(),
+			            ec.message().c_str());
+			return false;
 		}
+		std_fs::permissions(capture.path, std_fs::perms::owner_all, ec);
+
+		capture.next_index = {};
```

Comment:
This sets all indexes to -1, actually, and that's handled in `generate_capture_index` (-1 means a "reset"; we need to do that when we create a new capture dir, otherwise the highest indices are looked up on the first access per capture type, then we just use a running number for that type without any lookups).

Reply:
One a second thought, I could simplify it to remove the side-effect chain. Then we'd only try to create the capture dir on the (re-)init of the `[capture]` section and only looks up the indices then. What we'd lose is the automatic re-creation of the capture dir by `CAPTURE_CreateFile` if it was deleted... but honestly, that's no great loss. People should not delete the directories used by app while it's running 🤷🏻

Reply:
I think that latter case makes a lot of sense from a user perspective: the directory creation happens one time on startup (or when the conf options is updated).
This lets power-users easily swap things out if needed, and we'll be (deliberately) none the wiser.

Reply:
Okay, I'll simplify it then. It's a remnant of the old capture code, but it kinda grew on me... But yeah, it's just needlessly complicated.

Reply:
Done.